### ISSUE TITLE
Constant evaluation is not IEEE-754 compliant

### ISSUE THREAD

**Issue by LEI-Hongfaan** [dotnet/roslyn#64595]
**Steps to Reproduce**:

```
    Console.WriteLine(BitConverter.DoubleToInt64Bits(-double.NaN).ToString("X8"));
    var dn = double.NaN;
    Console.WriteLine(BitConverter.DoubleToInt64Bits(dn).ToString("X8"));
    Console.WriteLine(BitConverter.DoubleToInt64Bits(-dn).ToString("X8"));
```

**Expected Behavior**:

```
7FF8000000000000
FFF8000000000000
7FF8000000000000
```
No matter how double.NaN is defined, line 1 and line 3 should match.

**Actual Behavior**:

```
FFF8000000000000
FFF8000000000000
7FF8000000000000
```

**Comment by CyrusNajmabadi** [comment]
I'm not sure why these must match.  Doesn't IEEE define a broad set of values that are NaN?  Is there any guarantee that you'll get the same bit-values from that?  I'm not seeing anything in the C# spec that guarantees that.

**Comment by LEI-Hongfaan** [comment]
> I'm not sure why these must match. Doesn't IEEE define a broad set of values that are NaN? Is there any guarantee that you'll get the same bit-values from that?

Because compile-time and run-time result of -double.NaN should be the same. But compile-time gives wrong answer.
IEEE-754 says that negate is a bit-pattern operation which reverses the sign bit. It's somehow like copysign and behaves differently from the subtract operation which propagates NaN operand. Actually, IEEE-754 explicitly warns you about this: (-x) is not the same as (0 - x).

**Comment by CyrusNajmabadi** [comment]
> IEEE-754 says that negate is a bit-pattern operation which reverses the sign bit

Does it say this for NaN?

**Comment by CyrusNajmabadi** [comment]
> Because compile-time and run-time result of -double.NaN should be the same

Where is it specified that it must produce the same set of bits though?   That's what I'm asking.

**Comment by LEI-Hongfaan** [comment]
> Does it say this for NaN?

Yes if my understanding is correct. See IEEE Std 754-2019 § 5.5.1 Sign bit operations.

**Comment by tannergooding** [comment]
The TL;DR; is that Roslyn is not doing anything disallowed here.

-----------------------

5.5.1 Sign bit operations covers:
> For numeric x,
* copy(x) has the same representation as x, with no change of sign
* negate(x) has the same representation as x, but with the opposite sign
* abs(x) has the same representation as x, but with the sign 0 (positive)
* copySign(x, y) has the same representation as x, but with the sign of y if y is numeric and with unspecified sign if y is a NaN.
>
> For x a NaN, the results of the operations have the same representation as x (qNaN or sNaN)

Which only requires `qNaN` return `qNaN` and `sNaN` return `sNaN`. It does not strictly require the sign or payload to be preserved. Although many implementations do simply invert the sign to avoid the need to special-case the otherwise "one's complement" representation.

-----------------------

You should also always take into account `10.4 Literal meaning and value-changing optimizations`, as this section covers what types of transforms are considered "legal" because they preserve the original "intent" (or literal meaning) of the program.

> The following value-changing transformations, among others, preserve the literal meaning of the source
code:
> * ...
> * Changing the payload or sign bit of a quiet NaN.
> * ...

and you should take into account `11. Reproducible floating-point results`
> Users obtain the same floating-point numerical and reproducible status flag results, on all platforms
supporting such a language standard, by writing programs that:
> * ...
> and that
> * Do not use value-changing optimizations (see 10.4).
> * ...
> * Do not use signaling NaNs.
> * ...
> * Do not depend on quiet NaN propagation, payloads, or sign bits.
> * ...

The first point there is basically saying "don't use fast-math switches".
The second point is there because many programs and runtimes disable IEEE 754 floating-point exceptions and so `sNaN` is effectively the same as `qNaN`.
The third point is there because many programs and runtimes normalize `NaN` as an optimization, exactly what Roslyn is doing here.

**Comment by CyrusNajmabadi** [comment]
Closing out.  The impl is behaving in an entirely spec-compliant manner.

**Comment by LEI-Hongfaan** [comment]
> Where is it specified that it must produce the same set of bits though?

Nowhere. I don't think C# spec directly requires that. But the real problem is, assuming IEEE-754, current behavior implies that NaN in our implementation has two different bit representations even at compile-time solely. That's ridiculous and against existing [design](https://github.com/dotnet/roslyn/blob/e4667d4dc586d6bbf91ad32d7dd86f1a3971223b/src/Compilers/Core/Portable/ConstantValueSpecialized.cs#L20).

**Comment by CyrusNajmabadi** [comment]
> That's ridiculous 

I dont' understand what is ridiculous about it.  As you have stated, there is nowhere in either hte C# or IEEE specs that indicate the behavior you desire is required.  The compiler is spec-conformant here with both specifications, which seems completely sensible to me.

**Comment by LEI-Hongfaan** [comment]
> The TL;DR; is that Roslyn is not doing anything disallowed here.

Excuse me. My opinion: Roslyn is allowed to do such a thing because C#/Roslyn is not **_fully_** IEEE-754 compliant. I have never expected that. 
What you quote are requirements on non-interchange formats. Do you think System.Double is a non-interchange format? You can find the requirements on interchange formats (which of course are stricter) just some lines above. You should quote those lines.

You may decide a "Won't Fix". But I don't think you can say it's not a bug AND C#/Roslyn is IEEE-754 compliant simultaneously.

> The third point is there because many programs and runtimes normalize `NaN` as an optimization, exactly what Roslyn is doing here.

Roslyn is the compiler not the runtime, right? Doing so can make it compiles faster?

Let's see what other compilers (some are from M$) do.

C++
```C++
#include <ios>
#include <iostream>
#include <iomanip>

#include <limits>
#include <bit>
#include <cstdint>

using namespace std;

int main() {
    cout << hex << setw(16) << setfill('0') << bit_cast<uint64_t>(-numeric_limits<double>::quiet_NaN()) << endl;
    auto dn = numeric_limits<double>::quiet_NaN();
    cout << hex << setw(16) << setfill('0') << bit_cast<uint64_t>(dn) << endl;
    cout << hex << setw(16) << setfill('0') << bit_cast<uint64_t>(-dn) << endl;
    return 0;
}
```
MSVC
```
fff8000000000000
7ff8000000000000
fff8000000000000
```
Clang
```
fff8000000000000
7ff8000000000000
fff8000000000000
```

F#
```FSharp
open System

printfn "%s" (BitConverter.DoubleToInt64Bits(-Double.NaN).ToString("X16"))
let mutable dn = Double.NaN;
printfn "%s" (BitConverter.DoubleToInt64Bits(dn).ToString("X16"))
printfn "%s" (BitConverter.DoubleToInt64Bits(-dn).ToString("X16"))

printfn "========"
[<Literal>]
let anotherNaN = 0x7FF8000000000000LF

printfn "%s" (BitConverter.DoubleToInt64Bits(-anotherNaN).ToString("X16"))
let mutable dn1 = anotherNaN;
printfn "%s" (BitConverter.DoubleToInt64Bits(dn1).ToString("X16"))
printfn "%s" (BitConverter.DoubleToInt64Bits(-dn1).ToString("X16"))

```
```
7FF8000000000000
FFF8000000000000
7FF8000000000000
========
FFF8000000000000
7FF8000000000000
FFF8000000000000
```

Roslyn，你看看你，腰椎间盘都没你这么突出。As you can see, in all cases, line 1 and line 3 match. If you can find a compiler other than Roslyn to get line 1 and line 3 mismatch, please tell me.

Let me have the last question on this issue. Would you please hint me the location in source code where these constants get evaluated?

**Comment by tannergooding** [comment]
> What you quote are requirements on non-interchange formats. Do you think System.Double is a non-interchange format? You can find the requirements on interchange formats (which of course are stricter) just some lines above. You should quote those lines.

The allowances for `NaN` handling in `10.4` and `11` applies to all formats, interchange or otherwise.

**Comment by CyrusNajmabadi** [comment]
> Excuse me. My opinion: Roslyn is allowed to do such a thing because C#/Roslyn is not fully IEEE-754 compliant

Roslyn is an implementation of the c# language. Three main requirement it had is to implement that specification accurately. If it is doing so it is meeting its goal. 

So we would need actual links to them portions of its own spec (or any spec it references), to make a clear case why it is non-conformant. 

Currently it appears that 754 considers this behavior totally fine.  It is allowed to normalize -nan to the same bit pattern used by another nan. 

> Let's see what other compilers (some are from M$) do.

This is entirely irrelevant. Roslyn is not a c++ compiler. Nor does it's try to match the behavior of the compilers you mentioned. It is a c# (and vb) compiler and only cares about faithfully implementing the specifications for those languages.
